docs/cardano-monads.rst

```diff
@@ -0,0 +1,299 @@
+  and effects don't need to know about each other.
+
+* Flexibility. (Same as above, local use of effects).
+
+* Good run-time performance. Since in the end the entire monad transformer stack
+  was just `ReaderT` with a bunch of `IdentityT` on top (and occasional local
+  `StateT`), we enjoyed good run-time performance.
+
+* Conciseness. No boilerplate.
+
+Cons:
+
+* Bad run-time configurability. (Same as above)
+
+* Bad compile-time performance. Due to the way flattening works in Ether and due
+  to a GHC bug, the compile-time performance was devastating. Turning `-O2`
+  could mean hours of compilation and required up to 65 GIGABYTES of RAM
+  (ridiculous!). This was because GHC generated an exponential amount of
+  coercions (as evidenced by investigating .hi-files). Basically, I no longer
+  can recommend Ether to people as I have no good solution to this.
+
+Conclusion:
+
+* Migration to Ether allowed us to remove an immense amount of boilerplate,
+  modularize the code, and get good run-time performance. However, lack of
+  run-time configuability was quite inconvenient, and bad compile-time
+  performance marked this approach a no-go.
+
+The Modern Era
+--------------
+
+After we've realized what led to bad compile-time performance, I came up with an
+idea of `ExecMode`. Basically, we continued to use classes from `Ether`, but
+rather than having numerous `IdentityT` layers there was a single `newtype`
+wrapper around `ReaderT ModeEnv Production` at the bottom. This solved the
+compile-time performance issue completely at the cost of a moderate increase in
+boilerplate.
+
+However, FPComplete began to see `ether` as an Enemy of The State, and asked
+us to purge its remains. Now we were supposed to remove all our custom classes,
+replacing them with method records. Those records would go into a `ReaderT` and
+passed everywhere manually (as opposed to instance search). Instead of distinct
+`Ether.MonadReader` constraints, now we had to use `MonadReader ctx`, passing an
+annoying `ctx` parameter everywhere, and placing constraints on it. The final
+transformer stack is just `ReaderT ModeCtx Production`, and not even a newtype
+on top.
+
+Technically, we're still in the process of migration, as we haven't removed all
+of our custom classes yet. Just to clarify: in this section we'll discuss the
+current transitional state, and it's more painful than what was actually
+proposed by FPComplete.
+
+[Side note] As we realized that some of our custom effects were like `ReaderT`
+but did not require `local`, we replaced them with `reflection`. These effects
+basically were used to pass constant configuration to application components, so
+we used the dumb and unsafe `Given`-style reflection, avoiding the type-level
+complications of proper `Reifies`-style reflection. It turned out to be a great
+design choice: we've cut the amount of custom classes greatly, and the configs
+are now available even in class instances.
```

Comment:
Hm, this section is a bit strange. We didn't replace any `ReaderT`s with reflection. We only replaced configuration stuff, for which we never used `ReaderT`. At different stages we were using one of the following approaches for configuration:
* compile-time-only constants which are read using TH (`file-embed`);
* compile-time constants which can be modified in runtime using `unsafePerformIO`;
* eventually: reflection.

Reply:
Okay, I got an impression that some of the `ReaderT` were replaced too.

Reply:
It happened only once in WIP PR. I. e. there was a task to make something configurable, I started doing it using `ReaderT`, but then it was changed to use reflection within that PR, so `ReaderT` version wasn't finished and wasn't merged to master.